go/kbfs/libfuse/mounter.go

```diff
@@ -52,6 +54,17 @@ func (m *mounter) Mount() (err error) {
 }
 
 func fuseMountDir(dir string, platformParams PlatformParams) (*fuse.Conn, error) {
+	// Create mountdir directory on Linux
+	switch libkb.RuntimeGroup() {
+	case keybase1.RuntimeGroup_UNIXLIKE:
+		// Inherit permissions from containing directory and umask
+		err := os.MkdirAll(dir, os.ModeDir|os.ModePerm)
```

Comment:
It's the process that needs the directory, where would it be more appropriate? And this is by default in `$XDG_RUNTIME_DIR`, right? The lifecycle of the directory is already being managed as a tmpfs, and typically programs which need a runtime directory are responsible for creating it.

Reply:
The process also needs the keybase service, which it doesn't start itself.  It also needs a log file directory, which is doesn't create itself.  And as I've explained before, there are lots of situations (esp on macOS) where the default directory needs to be created with escalated permissions, and it doesn't make sense to me to build it into kbfsfuse for one platform but not the other.  That will make our lives harder down the road when debugging.

Reply:
(my opinion which is not backed by any real engineering experience)
I think it should create the logfile directory if it can't then, usually my programs on linux make directories on-demand if they need them as long as it's not going to conflict with another program. e.g., if a process wants a config file, it'll make `.config/`, `.config/process/`, `.config/process/config.json`. The service issue is something that should be handled by the process manager (which is systemd or `run_keybase` otherwise). 

It is true that for mounting filesystems with `mount` I usually need to `mkdir` first but I think that might be more historical than anything. At least, if you have the redirector, you shouldn't care too much about where this directory is.

I don't really see how this could become an issue basically, the Mac directory is created during install or on toggle kbfs right? I think since the operating systems already have different behavior for the mount, it's fine to make this nicer on Linux.

Reply:
Is `.ssh/` part of the environment that OpenSSH runs in? And yet, when running ssh-keygen for the first time, it will create that directory before creating files in it.

`systemctl --user enable --now keybase.service`, will create `.config/systemd/` in the process of creating the symlink `.config/systemd/user/default.target.wants/keybase.service`, rather than starting the service but logging a warning that it could not enable it long-term because "you didn't set up the environment beforehand".

I'm honestly not sure what this argument means. Either way, I guess distros could patch the service files to enable such a flag -- I simply don't understand why anyone would ever want to not do so.

...

It is worlds different to compare this to kbfsfuse not forking external services with a completely different lifetime, just because it depends on an **IPC socket**.

Reply:
You mean like the udisks daemon? That is directly responsible for simultaneously creating the mountpoint directory (in `/media` or `/run/media/$USER/`), and mounting your USB filesystem or whatever? And then deletes the mountpoint directory when the filesystem is unmounted?

The counterpoint, filesystems which do not have automatic setup and teardown of mountpoint directories, would be your root filesystem and anything else setup in `/etc/fstab`, which are meant to be constantly available from the day you partition the disks until the day years later that you retire the actual computer hardware itself.

Reply:
>That's part of what I mean when I'm talking about the environment -- it's something that kbfsfuse may not have control over.
Sure, but I think we should do the thing appropriate for the specific OS when on that OS. 

I feel that the vast majority of users are not going to want to manage the mountpoint themselves, and it's just another source of potential error in init scripts. Plus, I think it's cool if someone can just do `keybase service &; kbfsfuse &` and everything just works with sane defaults (I believe this more or less works, and I believe the service creates the log directory if it needs it, which is why kbfsfuse doesn't need to do it).

We can add a `--no-manage-mountdir` flag, I'm just having trouble coming up with why anyone would want to use it, which is why I don't think it should be the default.

I guess I don't see this as a high risk feature+it's convenient. What are the drawbacks to doing this @strib, other than inconsistency with OSX?

Reply:
Besides forking my (and maybe others') mental debugging model, the drawbacks/risks are:

1. When the dir cannot be created (due to permissions or whatever), we're shifting the error message from the `run_keybase` output to the logs, which the user might not know how to view or debug.  So if the user sets a custom mountpoint with a typo in it or something, it'll take some debugging to figure out what happened -- and a lot of that will flow through us.  This is a side effect of doing environmental things in the process itself, and is an example of the kind of unexpected things that can happen.
2. Potentially breaking some pre-existing custom setups out there that we don't know about.  For example, there could be some user/distro which depends on kbfs to fail mounting when the mountpoint doesn't exist.  Unlikely yes, but possible.  I just think we should be careful changing the default behavior of a years-old service.

These aren't dealbreakers of course.  @heronhaye if you decide to go ahead with this, it's fine, I didn't know you felt strongly about it.  I still somewhat disagree and I think we'll see some fallout from it, but if you're convinced the benefits outweigh the drawbacks, I'll withdraw my objections.

Reply:
Re 1: That is a good point I hadn't considered. I think what we could do is write to stdout if mounting (or other initialization) fails as well as log an error - then in the systemd case, journalctl will show the error (in addition to the kbfs logfile).

And I could be wrong but I think it'll also print to stdout in the terminal run_keybase was run on in the nonsystemd background case. 

Re 2: Hmm, it's possible. I think it could be more likely that people don't expect the directory to be deleted after KBFS shuts down, since that is new behavior in the successful case that run_keybase doesn't do. Maybe that should be behind a flag?

Reply:
Cool, if you can make any errors show up prominently in the logs and ideally on the terminal too, that would be a big help.

Good point about the directory deletion being new behavior.  If we keep this feature, I'd be ok with a flag or with just remembering whether we created the directory and only deleting it if we created it.

Reply:
>I'd be ok with a flag or with just remembering whether we created the directory and only deleting it if we created it.
That's a good idea, I'll do that.

Also re:
>Besides forking my (and maybe others') mental debugging model, 

What kind of bugs relating to this come up usually? Is there anything we can do to reduce them?

Reply:
> What kind of bugs relating to this come up usually? Is there anything we can do to reduce them?

Not sure I can give you too many specific examples, it's just a sense I have that issues involving the system environment are often hard to debug and take a lot of back-and-forth with the user.  The less confusion in our heads during that process, the better.  It's possible that in the long term, this change will lead to less confusion once we get used to it, I'm not sure.  Just thinking short term for now.

An example is what happens when someone's `$XDG_RUNTIME_DIR` is set to something unwritable or non-existent.  Normally I'd be able to see from the kbfsfuse command line call what the mount dir is, but now I'll have to look in a different place.